### PR TITLE
blockchain: proto migration 

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

migration of blockchain reactors to protobuf

Closes: #XXX

### CODE REVIEW

File: blockchain/v2/io.go [hunk 1]
```diff
@@ -59,7 +64,11 @@ func (sio *switchIO) sendStatusResponse(height int64, peerID p2p.ID) error {
 	if peer == nil {
 		return fmt.Errorf("peer not found")
 	}
-	msgBytes := cdc.MustMarshalBinaryBare(&bcStatusResponseMessage{Height: height})
+
+	msgBytes, err := bc.EncodeMsg(&bcproto.StatusResponse{Height: height})
```
Comment (marbar3778) [inline blockchain/v2/io.go:68]: @erikgrinaker is this correct? it's not filling Base

Comment (erikgrinaker) [reply blockchain/v2/io.go:68]: No, that doesn't seem right. Not sure what happened here, I'll add a separate PR.

File: blockchain/msgs.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+
+	bcproto "github.com/tendermint/tendermint/proto/blockchain"
+	"github.com/tendermint/tendermint/types"
+)
+
+const (
+	// NOTE: keep up to date with bcBlockResponseMessage
```
Comment (melekes) [inline blockchain/msgs.go:14]: can we write a test which checks that?

File: blockchain/msgs.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	BlockResponseMessagePrefixSize   = 4
+	BlockResponseMessageFieldKeySize = 1
+	MaxMsgSize                       = types.MaxBlockSizeBytes +
+		BlockResponseMessagePrefixSize +
+		BlockResponseMessageFieldKeySize
```
Comment (melekes) [inline blockchain/msgs.go:19]: how many fields are we talking about? 1?

File: blockchain/msgs.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+)
+
+// EncodeMsg encodes a Protobuf message, panicing on error.
```
Comment (melekes) [inline blockchain/msgs.go:22]:
```suggestion
// EncodeMsg encodes a Protobuf message.
```

File: blockchain/v1/reactor.go [hunk 5]
```diff
@@ -197,26 +200,44 @@ func (bcR *BlockchainReactor) AddPeer(peer p2p.Peer) {
...
-func (bcR *BlockchainReactor) sendBlockToPeer(msg *bcBlockRequestMessage,
+func (bcR *BlockchainReactor) sendBlockToPeer(msg *bcproto.BlockRequest,
 	src p2p.Peer) (queued bool) {
 
 	block := bcR.store.LoadBlock(msg.Height)
 	if block != nil {
-		msgBytes := cdc.MustMarshalBinaryBare(&bcBlockResponseMessage{Block: block})
+		pbbi, err := block.ToProto()
+		if err != nil {
+			bcR.Logger.Error("Could not send block message to peer: ", "err", err)
```
Comment (melekes) [inline blockchain/v1/reactor.go:210]:
```suggestion
			bcR.Logger.Error("Could not send block message to peer", "err", err)
```

File: blockchain/v1/reactor.go [hunk 6]
```diff
@@ -197,26 +200,44 @@ func (bcR *BlockchainReactor) AddPeer(peer p2p.Peer) {
...
+			return false
+		}
+		msgBytes, err := bc.EncodeMsg(&bcproto.BlockResponse{Block: pbbi})
+		if err != nil {
+			bcR.Logger.Error("unable to marshal msg: ", "err", err)
```
Comment (melekes) [inline blockchain/v1/reactor.go:215]:
```suggestion
			bcR.Logger.Error("unable to marshal msg", "err", err)
```

File: blockchain/v1/reactor.go [hunk 7]
```diff
@@ -197,26 +200,44 @@ func (bcR *BlockchainReactor) AddPeer(peer p2p.Peer) {
...
+		}
 		return src.TrySend(BlockchainChannel, msgBytes)
 	}
 
 	bcR.Logger.Info("peer asking for a block we don't have", "src", src, "height", msg.Height)
 
-	msgBytes := cdc.MustMarshalBinaryBare(&bcNoBlockResponseMessage{Height: msg.Height})
+	msgBytes, err := bc.EncodeMsg(&bcproto.NoBlockResponse{Height: msg.Height})
+	if err != nil {
+		bcR.Logger.Error("unable to marshal msg: ", "err", err)
```
Comment (melekes) [inline blockchain/v1/reactor.go:225]:
```suggestion
		bcR.Logger.Error("unable to marshal msg", "err", err)
```

File: blockchain/v1/reactor.go [hunk 8]
```diff
@@ -197,26 +200,44 @@ func (bcR *BlockchainReactor) AddPeer(peer p2p.Peer) {
...
 
-func (bcR *BlockchainReactor) sendStatusResponseToPeer(msg *bcStatusRequestMessage, src p2p.Peer) (queued bool) {
-	msgBytes := cdc.MustMarshalBinaryBare(&bcStatusResponseMessage{
+func (bcR *BlockchainReactor) sendStatusResponseToPeer(msg *bcproto.StatusRequest, src p2p.Peer) (queued bool) {
+	msgBytes, err := bc.EncodeMsg(&bcproto.StatusResponse{
 		Base:   bcR.store.Base(),
 		Height: bcR.store.Height(),
 	})
+	if err != nil {
+		bcR.Logger.Error("unable to marshal msg: ", "err", err)
```
Comment (melekes) [inline blockchain/v1/reactor.go:237]:
```suggestion
		bcR.Logger.Error("unable to marshal msg", "err", err)
```

File: blockchain/v1/reactor.go [hunk 9]
```diff
@@ -251,33 +272,38 @@ func (bcR *BlockchainReactor) Receive(chID byte, src p2p.Peer, msgBytes []byte)
...
 		// Send peer our state.
 		if queued := bcR.sendStatusResponseToPeer(msg, src); !queued {
 			// Unfortunately not queued since the queue is full.
 			bcR.Logger.Error("Could not send status message to peer", "src", src)
 		}
 
-	case *bcBlockResponseMessage:
+	case *bcproto.BlockResponse:
+		bi, _ := types.BlockFromProto(msg.Block)
+		// if err != nil {
```
Comment (melekes) [inline blockchain/v1/reactor.go:290]: ???

File: types/block.go [hunk 10]
```diff
@@ -835,6 +835,9 @@ func (commit *Commit) ValidateBasic() error {
 	if commit.Round < 0 {
 		return errors.New("negative Round")
 	}
+
+	// Check that commit.Height is greater than 1
+	// block 1 has commit of block 0 :?? so BlockID & CommitSigs are empty
```
Comment (melekes) [inline types/block.go:840]: what's the question here again?

File: blockchain/msgs.go (hunk 3 above)
Comment (marbar3778) [reply blockchain/msgs.go:19]: yes this would only be 1

**Pull request merged**